tests/test_build_nanoset_dataloader.py

```diff
@@ -0,0 +1,161 @@
+        # Create Nanosets
+        train_dataset, valid_dataset, test_dataset = NanosetBuilder(config).build()
+
+        assert isinstance(train_dataset, dataset_type)
+
+        if isinstance(train_dataset, Nanoset):
+            # Assert Nanoset is the same across all processes
+            assert_nanoset(train_dataset, parallel_context)
+
+        if isinstance(train_dataset, BlendedNanoset):
+            # Check the BlendedNanoset is composed of > 1 Nanoset
+            assert len(train_dataset.datasets) > 1
+            # Assert BlendedNanoset is the same across all processes
+            assert_blendednanoset(train_dataset, parallel_context)
+            # For each Nanoset in BlendedNanoset
+            for idx, dataset in enumerate(train_dataset.datasets):
+                # Check that each Nanoset of BlendedNanoset has more samples than the ones requested by BlendedNanoset
+                assert len(dataset) > get_max_value_by_group(
+                    train_dataset.dataset_index, train_dataset.dataset_sample_index, idx
+                )
+                # Assert Nanoset is the same across all processes
+                assert_nanoset(dataset, parallel_context)
+
+        # Create Dataloaders
+        dataloader = build_nanoset_dataloader(
+            train_dataset,
+            sequence_length=sequence_length,
+            parallel_context=parallel_context,
+            input_pp_rank=input_pp_rank,
+            output_pp_rank=output_pp_rank,
+            micro_batch_size=MICRO_BATCH_SIZE,
+            dataloader_num_workers=0,
+            dataloader_drop_last=True,
+        )
+
+        # Check a batch produced by the Dataloader
+        batch = next(iter(dataloader))
+        assert_batch_dataloader(
+            batch=batch,
+            parallel_context=parallel_context,
+            micro_batch_size=MICRO_BATCH_SIZE,
+            sequence_length=sequence_length,
+        )
+
+        # Recover from failures
+        SKIPED_BATCHES = 20
+        dataloader = iter(dataloader)
+        for _ in range(SKIPED_BATCHES + 1):  # In order to compare with the first batch of the recovered DataLoader
+            batch = next(dataloader)
+
+        recovered_dataloader = build_nanoset_dataloader(
+            train_dataset,
+            sequence_length=sequence_length,
+            parallel_context=parallel_context,
+            input_pp_rank=input_pp_rank,
+            output_pp_rank=output_pp_rank,
+            micro_batch_size=MICRO_BATCH_SIZE,
+            dataloader_num_workers=0,
+            dataloader_drop_last=True,
+            # NOTE The dataloader serves batches of micro_batch_size despite of batch_accumulation_per_replica
+            consumed_train_samples=SKIPED_BATCHES * MICRO_BATCH_SIZE * parallel_context.dp_pg.size(),
+        )
+
+        recovered_first_batch = next(iter(recovered_dataloader))
+
+        assert compute_batch_hash(batch) == compute_batch_hash(recovered_first_batch)
```

Comment:
We could move this to a separate test (making each test only test a specific expected behavior).

Reply:
I have divided the tests: Now we have `test_build_nanoset_dataloader`, which checks that we correctly build the datasets and DataLoaders in each process, and `test_recover_nanoset_dataloader`, which checks that when restarting training, we are able to recover the state prior to the interruption. In the case of `test_recover_nanoset_dataloader`, it is necessary to `skip batches > train steps`.